chef/cookbooks/ceph/providers/client.rb

```diff
@@ -61,23 +61,23 @@ def get_caps(ceph_conf, admin_keyring, keyname)
 end
 
 def auth_set_key(ceph_conf, admin_keyring, keyname, caps)
+  raise "No caps defined for ceph_client resource!" if caps.empty?
   # try to add the key
   caps_str = caps.map { |k, v| "#{k} '#{v}'" }.join(" ")
   cmd = "ceph -k #{admin_keyring} -c #{ceph_conf} auth get-or-create #{keyname} #{caps_str}"
   get_or_create = Mixlib::ShellOut.new(cmd)
   get_or_create.run_command
-  unless get_or_create.stderr.scan(/EINVAL.*but cap.*does not match/).empty?
-    Chef::Log.info("Deleting old key with incorrect caps")
-    # don't remove custom existing caps for osd pools
-    caps_osd = get_caps(ceph_conf, admin_keyring, keyname)["osd"] + "," + caps["osd"]
-    caps["osd"] = caps_osd.split(",").collect(&:strip).uniq.join(",")
-    caps_str = caps.map { |k, v| "#{k} '#{v}'" }.join(" ")
-    cmd = "ceph -k #{admin_keyring} -c #{ceph_conf} auth get-or-create #{keyname} #{caps_str}"
-    # delete an old key if it exists and is wrong
-    Mixlib::ShellOut.new("ceph -k #{admin_keyring} -c #{ceph_conf} auth del #{keyname}").run_command
-    # try to create again
-    get_or_create = Mixlib::ShellOut.new(cmd)
-    get_or_create.run_command
-  end
-  get_or_create.error!
+  return get_or_create.error! unless get_or_create.stderr.scan(/EINVAL.*but cap.*does not match/)
+  Chef::Log.info("Updating incorrect caps for #{keyname}")
+  # don't remove custom existing caps for osd pools
+  cur_caps = get_caps(ceph_conf, admin_keyring, keyname)
+  caps_osd = caps["osd"]
+  caps_osd.insert(0, cur_caps["osd"] + ",") unless cur_caps.empty?
+  caps["osd"] = caps_osd.split(",").collect(&:strip).uniq.join(", ")
```

Comment:
As far as I understand it, the uniq method uses the hashes of the list object to uniquify the list, and sorting it beforehand is not necessary.